File: ios/LoginScreen/LoginScreen/LoginViewModel.swift
import Foundation
import Combine

/// ViewModel for the login screen following MVVM architecture.
/// Manages UI state, handles login logic, validation, and state updates.
/// Observes network state and implements lockout mechanism after 3 failures.
@MainActor
class LoginViewModel: ObservableObject {
    /// Published state that triggers UI updates when changed.
    @Published var state = LoginState()
    
    private let authService: AuthServiceProtocol
    private let networkMonitor: NetworkMonitorProtocol
    private let tokenStorage: TokenStorageProtocol
    private var cancellables = Set<AnyCancellable>()
    
    /// Initializes the ViewModel with required dependencies.
    ///
    /// - Parameters:
    ///   - authService: Service for authentication operations
    ///   - networkMonitor: Monitor for network connectivity state
    ///   - tokenStorage: Storage for persisting authentication tokens
    init(
        authService: AuthServiceProtocol,
        networkMonitor: NetworkMonitorProtocol,
        tokenStorage: TokenStorageProtocol
    ) {
        self.authService = authService
        self.networkMonitor = networkMonitor
        self.tokenStorage = tokenStorage
        
        setupNetworkMonitoring()
        setupValidation()
        loadSavedToken()
    }
    
    /// Updates the username in the UI state.
    /// - Parameter username: The new username value
    func updateUsername(_ username: String) {
        state.username = username
    }
    
    /// Updates the password in the UI state.
    /// - Parameter password: The new password value
    func updatePassword(_ password: String) {
        state.password = password
    }
    
    /// Updates the remember me checkbox state.
    /// - Parameter rememberMe: The new remember me value
    func updateRememberMe(_ rememberMe: Bool) {
        state.rememberMe = rememberMe
    }
    
    /// Attempts to login with current username and password.
    /// Validates state (not locked out, online, button enabled) before proceeding.
    /// Updates failure count and locks account after 3 failures.
    func login() {
        guard !state.isLockedOut, !state.isOffline, state.isButtonEnabled else {
            return
        }
        
        state.isLoading = true
        state.errorMessage = nil
        
        Task {
            do {
                let token = try await authService.login(
                    username: state.username,
                    password: state.password
                )
                
                // Save token if remember me is enabled
                if state.rememberMe {
                    await tokenStorage.saveToken(token)
                } else {
                    await tokenStorage.clearToken()
                }
                
                state.isLoading = false
                state.errorMessage = nil
                state.failureCount = 0
                state.navigationEvent = .navigateToHome
            } catch {
                let newFailureCount = state.failureCount + 1
                let isLockedOut = newFailureCount >= 3
                
                state.isLoading = false
                state.errorMessage = error.localizedDescription
                state.failureCount = newFailureCount
                state.isLockedOut = isLockedOut
            }
        }
    }
    
    /// Clears the current error message from the UI state.
    func clearError() {
        state.errorMessage = nil
    }
    
    /// Clears the navigation event after it has been consumed.
    func clearNavigationEvent() {
        state.navigationEvent = nil
    }
    
    /// Sets up network monitoring to update offline state reactively.
    private func setupNetworkMonitoring() {
        networkMonitor.isOnline
            .receive(on: DispatchQueue.main)
            .map { !$0 }
            .assign(to: \.state.isOffline, on: self)
            .store(in: &cancellables)
    }
    
    /// Sets up form validation to enable/disable login button reactively.
    /// Button is enabled when: fields are not empty, not loading, not locked out, and online.
    private func setupValidation() {
        $state
            .combineLatest(networkMonitor.isOnline)
            .map { state, isOnline in
                !state.username.isEmpty &&
                !state.password.isEmpty &&
                !state.isLoading &&
                !state.isLockedOut &&
                isOnline
            }
            .assign(to: \.state.isButtonEnabled, on: self)
            .store(in: &cancellables)
    }
    
    /// Loads saved token on initialization and sets remember me if token exists.
    private func loadSavedToken() {
        Task {
            let savedToken = await tokenStorage.getToken()
            if savedToken != nil {
                state.rememberMe = true
            }
        }
    }
}

/// Protocol for token storage operations.
/// Provides abstraction for persisting and retrieving authentication tokens.
protocol TokenStorageProtocol {
    /// Saves an authentication token.
    /// - Parameter token: The token to save
    func saveToken(_ token: String) async
    
    /// Retrieves the saved authentication token.
    /// - Returns: The saved token, or nil if no token exists
    func getToken() async -> String?
    
    /// Clears the saved authentication token.
    func clearToken() async
}

/// iOS implementation of TokenStorageProtocol using UserDefaults.
class TokenStorage: TokenStorageProtocol {
    private let userDefaults = UserDefaults.standard
    private let tokenKey = "auth_token"
    
    func saveToken(_ token: String) async {
        userDefaults.set(token, forKey: tokenKey)
    }
    
    func getToken() async -> String? {
        return userDefaults.string(forKey: tokenKey)
    }
    
    func clearToken() async {
        userDefaults.removeObject(forKey: tokenKey)
    }
}

File: android/app/src/main/java/com/loginscreen/LoginViewModel.kt
package com.loginscreen

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * ViewModel for the login screen following MVVM architecture.
 * Manages UI state, handles login logic, validation, and state updates.
 * Observes network state and implements lockout mechanism after 3 failures.
 *
 * @param authRepository Repository for authentication operations
 * @param networkMonitor Monitor for network connectivity state
 * @param tokenStorage Storage for persisting authentication tokens
 */
class LoginViewModel(
    private val authRepository: AuthRepository,
    private val networkMonitor: NetworkMonitor,
    private val tokenStorage: TokenStorage
) : ViewModel() {

    private val _uiState = MutableStateFlow(LoginUiState())
    val uiState: StateFlow<LoginUiState> = _uiState.asStateFlow()

    init {
        // Observe network state and update offline status
        viewModelScope.launch {
            networkMonitor.isOnline.collect { isOnline ->
                _uiState.update { it.copy(isOffline = !isOnline) }
            }
        }

        // Observe form validation and network state to enable/disable login button
        viewModelScope.launch {
            combine(
                _uiState,
                networkMonitor.isOnline
            ) { state, isOnline ->
                val isValid = state.username.isNotBlank() &&
                        state.password.isNotBlank() &&
                        !state.isLoading &&
                        !state.isLockedOut &&
                        isOnline
                state.copy(isButtonEnabled = isValid)
            }.collect { updatedState ->
                _uiState.update { updatedState }
            }
        }

        // Load saved token if remember me was enabled
        loadSavedToken()
    }

    /**
     * Updates the username in the UI state.
     * @param username The new username value
     */
    fun updateUsername(username: String) {
        _uiState.update { it.copy(username = username) }
    }

    /**
     * Updates the password in the UI state.
     * @param password The new password value
     */
    fun updatePassword(password: String) {
        _uiState.update { it.copy(password = password) }
    }

    /**
     * Updates the remember me checkbox state.
     * @param rememberMe The new remember me value
     */
    fun updateRememberMe(rememberMe: Boolean) {
        _uiState.update { it.copy(rememberMe = rememberMe) }
    }

    /**
     * Attempts to login with current username and password.
     * Validates state (not locked out, online, button enabled) before proceeding.
     * Updates failure count and locks account after 3 failures.
     */
    fun login() {
        val currentState = _uiState.value
        if (currentState.isLockedOut || currentState.isOffline || !currentState.isButtonEnabled) {
            return
        }

        viewModelScope.launch {
            _uiState.update {
                it.copy(
                    isLoading = true,
                    errorMessage = null
                )
            }

            val result = authRepository.login(
                currentState.username,
                currentState.password
            )

            _uiState.update { state ->
                result.fold(
                    onSuccess = { token ->
                        // Save token if remember me is enabled
                        if (state.rememberMe) {
                            tokenStorage.saveToken(token)
                        } else {
                            tokenStorage.clearToken()
                        }

                        state.copy(
                            isLoading = false,
                            errorMessage = null,
                            failureCount = 0,
                            navigationEvent = NavigationEvent.NavigateToHome
                        )
                    },
                    onFailure = { error ->
                        val newFailureCount = state.failureCount + 1
                        val isLockedOut = newFailureCount >= 3

                        state.copy(
                            isLoading = false,
                            errorMessage = error.message ?: "Login failed",
                            failureCount = newFailureCount,
                            isLockedOut = isLockedOut
                        )
                    }
                )
            }
        }
    }

    /**
     * Clears the current error message from the UI state.
     */
    fun clearError() {
        _uiState.update { it.copy(errorMessage = null) }
    }

    /**
     * Clears the navigation event after it has been consumed.
     */
    fun clearNavigationEvent() {
        _uiState.update { it.clearNavigationEvent() }
    }

    /**
     * Loads saved token on initialization and sets remember me if token exists.
     */
    private fun loadSavedToken() {
        viewModelScope.launch {
            val savedToken = tokenStorage.getToken()
            if (savedToken != null) {
                _uiState.update { it.copy(rememberMe = true) }
            }
        }
    }
}

/**
 * Interface for token storage operations.
 * Provides abstraction for persisting and retrieving authentication tokens.
 */
interface TokenStorage {
    /**
     * Saves an authentication token.
     * @param token The token to save
     */
    suspend fun saveToken(token: String)

    /**
     * Retrieves the saved authentication token.
     * @return The saved token, or null if no token exists
     */
    suspend fun getToken(): String?

    /**
     * Clears the saved authentication token.
     */
    suspend fun clearToken()
}

/**
 * Android implementation of TokenStorage using SharedPreferences.
 */
class TokenStorageImpl(private val context: android.content.Context) : TokenStorage {
    private val prefs = context.getSharedPreferences("auth_prefs", android.content.Context.MODE_PRIVATE)

    override suspend fun saveToken(token: String) {
        prefs.edit().putString("auth_token", token).apply()
    }

    override suspend fun getToken(): String? {
        return prefs.getString("auth_token", null)
    }

    override suspend fun clearToken() {
        prefs.edit().remove("auth_token").apply()
    }
}

